tests/msc3391_test.go

```diff
@@ -0,0 +1,274 @@
+func createAccountData(t *testing.T, c *client.CSAPI, roomID *string) {
+	// a function to check that the content of a user or account data object
+	// matches our test content.
+	checkAccountData := func(r gjson.Result) bool {
+		// Only listen for our test type
+		if r.Get("type").Str != testAccountDataType {
+			return false
+		}
+		content := r.Get("content")
+
+		// Ensure the content of this account data type is as we expect
+		return match.JSONDeepEqual([]byte(content.Raw), testAccountDataContent)
+	}
+
+	// Retrieve a sync token for this user
+	_, nextBatchToken := c.MustSync(
+		t,
+		client.SyncReq{},
+	)
+
+	// Set and check the account data
+	if roomID != nil {
+		// Create room account data
+		c.SetRoomAccountData(t, *roomID, testAccountDataType, testAccountDataContent)
+
+		// Wait for the account data to appear down /sync
+		c.MustSyncUntil(
+			t,
+			client.SyncReq{
+				Since: nextBatchToken,
+			},
+			client.SyncRoomAccountDataHas(*roomID, checkAccountData),
+		)
+
+		// Also check the account data content by querying the appropriate endpoint
+		res := c.GetRoomAccountData(t, *roomID, testAccountDataType)
+		must.MatchResponse(t, res, match.HTTPResponse{
+			JSON: []match.JSON{
+				func(body []byte) error {
+					if !match.JSONDeepEqual(body, testAccountDataContent) {
+						return fmt.Errorf(
+							"Expected %s for room account data content when, got '%s'",
+							testAccountDataType,
+							string(body),
+						)
+					}
+
+					return nil
+				},
+			},
+		})
+	} else {
+		// Create user account data
+		c.SetGlobalAccountData(t, testAccountDataType, testAccountDataContent)
+
+		// Wait for the account data to appear down /sync
+		c.MustSyncUntil(
+			t,
+			client.SyncReq{
+				Since: nextBatchToken,
+			},
+			client.SyncGlobalAccountDataHas(checkAccountData),
+		)
+
+		// Also check the account data content by querying the appropriate endpoint
+		res := c.GetGlobalAccountData(t, testAccountDataType)
+		must.MatchResponse(t, res, match.HTTPResponse{
+			JSON: []match.JSON{
+				func(body []byte) error {
+					if !match.JSONDeepEqual(body, testAccountDataContent) {
+						return fmt.Errorf(
+							"Expected %s for room account data content when, got '%s'",
+							testAccountDataType,
+							string(body),
+						)
+					}
+
+					return nil
+				},
+			},
+		})
+	}
+}
+
+// deleteAccountData removes account data for a user or room.
+//
+// If viaDelete is true, a request is made to the DELETE endpoint for user or
+// room account data. Otherwise, the PUT method is used with an empty content
+// dictionary instead. MSC3391 specifies that a PUT with an empty content body
+// is functionally equivalent to deleting an account data type directly.
+//
+// If roomID is not nil, room account data for the given room ID will be removed.
+// Otherwise, account data from the user will be removed instead.
+func deleteAccountData(t *testing.T, c *client.CSAPI, viaDelete bool, roomID *string) {
```

Comment:
Similar question here of if we can separate the global / room functionality.

Reply:
Good point. I've separated out the various test methods into user and room variants in 58913da.